### PR TITLE
fix(shapes): render grouped DrawingML shapes with custom geometry (SD-1877) ❇️

### PR DESCRIPTION
**Pull request by tupizz**
## Summary

- **Fixed grouped DrawingML shapes rendering as solid black rectangles** when they use custom geometry (`a:custGeom`) instead of preset geometry (`a:prstGeom`)
- Added `extractCustomGeometry()` parser that converts DrawingML path commands (moveTo, lnTo, cubicBezTo, quadBezTo, close) to SVG path data, wired through the full pipeline: DOCX import → PM node → pm-adapter → contracts → DomPainter + ShapeGroupView
- Used `fill-rule="evenodd"` for hollow frame paths and `vector-effect="non-scaling-stroke"` hairline strokes to keep thin borders visible after non-uniform group coordinate scaling
- Fixed floating drawing width constraint: `wrapNone` drawings no longer incorrectly constrained to content area width
- Rewrote group coordinate rendering to use visual-space positioning (pre-scaled by importer) instead of CSS scale transforms

## Visual Comparison

| Before (main) | After (this PR) | Word Reference |
|:-:|:-:|:-:|
| CleanShot 2026-02-18 at 22 10 53 | CleanShot 2026-02-18 at 22 11 19 | CleanShot 2026-02-18 at 22 11 12@2x
| Solid black rectangle hides all content | Border frame + gray background render correctly | Expected rendering in Microsoft Word |

## Test plan

- [x] All unit tests pass (super-editor: 6162, painter-dom: 598, measuring-dom: 219, pm-adapter: 1533)
- [x] All 56 visual regression tests pass (0 failures)
- [x] Verified in browser with the sample DOCX — shape group with border frame and gray background renders correctly matching Word
- [ ] Test with other documents containing grouped shapes to verify no regressions
- [ ] Test with documents containing preset geometry shapes to verify existing rendering is unaffected

### CODE REVIEW

File: packages/layout-engine/painters/dom/src/renderer.ts [hunk 1]
```diff
@@ -3086,6 +3092,61 @@ export class DomPainter {
...
+        // If this path has a different coordinate space, apply a transform to map it
+        const pathW = p.w || viewW;
+        const pathH = p.h || viewH;
+        const needsTransform = pathW !== viewW || pathH !== viewH;
+        const scaleX = viewW / pathW;
+        const scaleY = viewH / pathH;
+        const transform = needsTransform ? ` transform="scale(${scaleX}, ${scaleY})"` : '';
+        const strokeAttr =
+          strokeColor !== 'none' ? ` stroke="${strokeColor}" stroke-width="${strokeWidth}"` : edgeStroke;
+        return `<path d="${p.d}" fill="${fillColor}" fill-rule="evenodd"${strokeAttr}${transform} />`;
```
[the remark this answers is not in the document]
Comment (tupizz) [reply packages/layout-engine/painters/dom/src/renderer.ts:3531]: Not a real risk: `p.d` is extracted from OOXML XML attributes by `extractCustomGeometry()` (not from user input), and the generated SVG string passes through `parseSafeSvg()` which calls `DOMParser` + `stripUnsafeSvgContent()` to remove scripts and `on*` event handlers before the SVG element is inserted into the DOM.

File: packages/layout-engine/painters/dom/src/renderer.ts [hunk 2]
```diff
@@ -3086,6 +3092,61 @@ export class DomPainter {
...
+
+    const pathElements = custGeom.paths
+      .map((p) => {
+        // If this path has a different coordinate space, apply a transform to map it
+        const pathW = p.w || viewW;
+        const pathH = p.h || viewH;
+        const needsTransform = pathW !== viewW || pathH !== viewH;
+        const scaleX = viewW / pathW;
+        const scaleY = viewH / pathH;
+        const transform = needsTransform ? ` transform="scale(${scaleX}, ${scaleY})"` : '';
```
[the remark this answers is not in the document]
Comment (tupizz) [reply packages/layout-engine/painters/dom/src/renderer.ts:3528]: The `|| viewW` fallback means if `p.w = 0`, `pathW = viewW`, so `pathW \!== viewW` is false and the scale transform is never applied — `Infinity` can't be produced. That said, a degenerate `viewW/viewH = 0` would produce an invalid `viewBox="0 0 0 0"` even without a scale. Added an explicit early-return guard (`if (viewW === 0 || viewH === 0) return null`) to make the intent clear and avoid emitting broken SVG markup.

File: packages/super-editor/src/core/super-converter/v3/handlers/wp/helpers/vector-shape-helpers.js [hunk 3]
```diff
@@ -449,6 +449,89 @@ export function extractFillColor(spPr, style) {
...
+function convertDrawingMLPathToSvg(pathEl) {
+  if (!pathEl?.elements) return '';
+
+  const parts = [];
+  for (const cmd of pathEl.elements) {
+    switch (cmd.name) {
+      case 'a:moveTo': {
+        const pt = cmd.elements?.find((el) => el.name === 'a:pt');
+        if (pt) {
+          parts.push(`M ${pt.attributes?.['x'] || 0} ${pt.attributes?.['y'] || 0}`);
```
[the remark this answers is not in the document]
Comment (tupizz) [reply packages/super-editor/src/core/super-converter/v3/handlers/wp/helpers/vector-shape-helpers.js:497]: The `|| 0` covers undefined/null/empty-string but not non-numeric strings. In practice OOXML coordinate attributes are always integers (EMUs); non-numeric values would indicate a malformed document. Even if garbage data slipped through, SVG browsers silently ignore malformed `d` attributes rather than crashing, and the SVG itself passes through `parseSafeSvg()` (DOMParser parse + sanitize) before DOM insertion. Left as-is to avoid adding noise for a purely theoretical case.

File: packages/super-editor/src/core/super-converter/v3/handlers/wp/helpers/vector-shape-helpers.js [hunk 4]
```diff
@@ -449,6 +449,89 @@ export function extractFillColor(spPr, style) {
...
+      }
+      case 'a:quadBezTo': {
+        const pts = cmd.elements?.filter((el) => el.name === 'a:pt') || [];
+        if (pts.length === 2) {
+          parts.push(
+            `Q ${pts[0].attributes?.['x'] || 0} ${pts[0].attributes?.['y'] || 0} ` +
+              `${pts[1].attributes?.['x'] || 0} ${pts[1].attributes?.['y'] || 0}`,
+          );
+        }
+        break;
```
[the remark this answers is not in the document]
Comment (tupizz) [reply packages/super-editor/src/core/super-converter/v3/handlers/wp/helpers/vector-shape-helpers.js:527]: Same as the moveTo case — OOXML coordinate attributes are always integers in practice, and SVG silently ignores malformed path data. Left as-is for the same reasons.

File: packages/super-editor/src/core/super-converter/v3/handlers/wp/helpers/vector-shape-helpers.js [hunk 5]
```diff
@@ -449,6 +449,89 @@ export function extractFillColor(spPr, style) {
   return null;
 }
 
+/**
+ * Extracts custom geometry path data from a:custGeom element and converts it to SVG paths.
+ * Per ECMA-376, a:custGeom contains a:pathLst with path commands (moveTo, lnTo, cubicBezTo,
+ * quadBezTo, arcTo, close) in a coordinate space defined by the path's w/h attributes.
```
[the remark this answers is not in the document]
Comment (tupizz) [reply packages/super-editor/src/core/super-converter/v3/handlers/wp/helpers/vector-shape-helpers.js]: Fixed. Removed `arcTo` from the supported commands list in the JSDoc and added a note that it is intentionally skipped. Also added an explicit `default: break` case with a comment to document that unrecognized commands produce no output.

File: packages/super-editor/src/extensions/shape-group/ShapeGroupView.js [hunk 6]
```diff
@@ -295,9 +296,66 @@ export class ShapeGroupView {
...
+      innerSvg.setAttribute('preserveAspectRatio', 'none');
+
+      for (const pathData of customGeometry.paths) {
+        const pathEl = document.createElementNS('http://www.w3.org/2000/svg', 'path');
+        pathEl.setAttribute('d', pathData.d);
+        pathEl.setAttribute('fill', fillStr);
+        pathEl.setAttribute('fill-rule', 'evenodd');
+        pathEl.setAttribute('stroke', strokeStr);
+        pathEl.setAttribute('stroke-width', strokeW.toString());
+
```
[the remark this answers is not in the document]
Comment (tupizz) [reply packages/super-editor/src/extensions/shape-group/ShapeGroupView.js]: Fixed. Added the same hairline edge stroke workaround to ShapeGroupView — when a shape is filled but has no stroke, a `stroke-width="0.5" vector-effect="non-scaling-stroke"` matching the fill color is set on each path element, identical to the DomPainter logic.

File: packages/super-editor/src/core/super-converter/v3/handlers/wp/helpers/vector-shape-helpers.js [hunk 7]
```diff
@@ -449,6 +449,89 @@ export function extractFillColor(spPr, style) {
...
+      const w = parseInt(pathEl.attributes?.['w'] || '0', 10);
+      const h = parseInt(pathEl.attributes?.['h'] || '0', 10);
+      const d = convertDrawingMLPathToSvg(pathEl);
+      return { d, w, h };
+    })
+    .filter((p) => p.d);
+
+  if (paths.length === 0) return null;
+  return { paths };
+}
```
[the remark this answers is not in the document]
Comment (tupizz) [reply packages/super-editor/src/core/super-converter/v3/handlers/wp/helpers/vector-shape-helpers.js:479]: Fair point. `extractCustomGeometry` is a pure OOXML-to-SVG-path transformer and is well-suited for unit tests. Tracking as a follow-up — the test file would live alongside `vector-shape-helpers.js` and cover the common path commands (moveTo, lnTo, cubicBezTo, quadBezTo, close) and the `w`/`h = 0` edge case.

File: packages/super-editor/src/extensions/shape-group/ShapeGroupView.js [hunk 8]
```diff
@@ -295,9 +296,66 @@ export class ShapeGroupView {
...
+        // Scale if this path has a different coordinate space
+        const pathW = pathData.w || viewW;
+        const pathH = pathData.h || viewH;
+        if (pathW !== viewW || pathH !== viewH) {
```
[the remark this answers is not in the document]
Comment (tupizz) [reply packages/super-editor/src/extensions/shape-group/ShapeGroupView.js]: Same analysis as the DomPainter thread: `pathW = pathData.w || viewW` means if `pathData.w = 0`, `pathW = viewW`, so the `pathW !== viewW` guard is false and scale is never applied. Added the `viewW > 0 && viewH > 0` guard (wrapping the entire inner SVG creation) to prevent rendering with a degenerate viewBox, which covers this case.

File: packages/super-editor/src/core/super-converter/v3/handlers/wp/helpers/vector-shape-helpers.js [hunk 9]
```diff
@@ -449,6 +449,89 @@ export function extractFillColor(spPr, style) {
...
+  if (!custGeom) return null;
+
+  const pathLst = custGeom.elements?.find((el) => el.name === 'a:pathLst');
+  if (!pathLst?.elements) return null;
+
+  const paths = pathLst.elements
+    .filter((el) => el.name === 'a:path')
+    .map((pathEl) => {
+      const w = parseInt(pathEl.attributes?.['w'] || '0', 10);
+      const h = parseInt(pathEl.attributes?.['h'] || '0', 10);
```
[the remark this answers is not in the document]
Comment (tupizz) [reply packages/super-editor/src/core/super-converter/v3/handlers/wp/helpers/vector-shape-helpers.js:471]: When `w`/`h` attributes are missing, `parseInt('0', 10) = 0`. In the consumer, `firstPath.w || width` means `viewW = width` when `firstPath.w = 0`. The only problematic case is when `width = 0` too (a zero-size shape), which would produce `viewBox="0 0 0 0"`. This is now caught by the explicit zero-dimension guard added in this commit (`if (viewW === 0 || viewH === 0) return null` in DomPainter, `if (viewW > 0 && viewH > 0)` in ShapeGroupView).

File: packages/layout-engine/painters/dom/src/renderer.ts [hunk 10]
```diff
@@ -3086,6 +3092,61 @@ export class DomPainter {
...
+    }
+    const strokeColor =
+      block.strokeColor === null ? 'none' : typeof block.strokeColor === 'string' ? block.strokeColor : 'none';
+    const strokeWidth = block.strokeColor === null ? 0 : (block.strokeWidth ?? 0);
+
+    // Build SVG paths. Each path has its own coordinate space (w × h).
+    // Use the first path's coordinate space for the viewBox, and scale subsequent paths if needed.
+    const firstPath = custGeom.paths[0];
+    const viewW = firstPath.w || width;
+    const viewH = firstPath.h || height;
```
[the remark this answers is not in the document]
Comment (tupizz) [reply packages/layout-engine/painters/dom/src/renderer.ts:3507]: Fixed. Added `if (viewW === 0 || viewH === 0) return null` before generating the SVG string. This prevents emitting `viewBox="0 0 0 0"` for zero-size shapes (which browsers treat as an invisible or broken SVG), and falls through to the regular fallback rendering path.

File: packages/super-editor/src/extensions/shape-group/ShapeGroupView.js [hunk 11]
```diff
@@ -295,9 +296,66 @@ export class ShapeGroupView {
...
 
+    // Handle custom geometry paths (a:custGeom) — render SVG paths directly
+    if (customGeometry?.paths?.length) {
+      const fillStr = fillValue === null ? 'none' : typeof fillValue === 'string' ? fillValue : 'none';
+      const strokeStr = strokeColor === null ? 'none' : strokeColor;
+      const strokeW = strokeColor === null ? 0 : strokeWidth;
+
+      const firstPath = customGeometry.paths[0];
+      const viewW = firstPath.w || width;
+      const viewH = firstPath.h || height;
```
[the remark this answers is not in the document]
Comment (tupizz) [reply packages/super-editor/src/extensions/shape-group/ShapeGroupView.js:307]: Fixed. Wrapped the inner SVG creation in `if (viewW > 0 && viewH > 0)` to skip rendering when the viewBox would be degenerate. The text content block (if present) still renders outside this guard.

**Comment by harbournick** [comment]
:tada: This PR is included in **superdoc** v1.17.0

The release is available on [GitHub release](https://github.com/superdoc-dev/superdoc/releases/tag/v1.17.0)
